design-documents/service-isolation/integration-tests.md

```diff
@@ -0,0 +1,46 @@
+# Integration Tests
+
+## Problem Statement
+
+Integration tests run in the same PHP process as Magento application.
+A test from one module may depend on a data fixture or a model from another module.
+In case of **single-area setup** or setup with **different application areas served by different instances**, integration tests and their dependencies may be deployed on **different instances** and, as a result, such tests won't be able to resolve their dependencies.
+
+The question: should integration testing framework support distributed deployment or keep it compatible to monolithic deployment only?
+
+Take into account future support of microservices architecture.
+
+## Purpose of Integration Tests in Magento
+
+Before deciding whether we should drop support of integration tests for distributed systems, let's revisit purpose of integration tests in Magento:
+1. validate extension points of a component (potential integration with other components)
+2. validate integration with other existing components. Integration tests don't use mocks (usually)
+3. validate integration with the environment. Integration tests run on installed application, with real DB, etc
+
+
+## Solution
+
+From the purpose of integration tests described above, 
+- 2 and 3 can and should be covered by functional tests as they validate real scenarios and so bring more value
+- 1 still makes sense for integration tests, but only in scope of one instance (single-area application or service) because this is where in-process extension points are supported  
+
+Based on this, it makes sense to run integration tests for a single-area application instance (or for a service in the future).
+
+It should be taken into account that in case of multiple single-area application, DB is shared between all of them.
+
+Changes to integration tests:
+1. Break down tests by module (according to new module structure).
+  1. Testing package declares which Magento package it depends on
+  1. Testing packages are deployed on the same instances their dependencies are deployed. Deployment can be managed in the same way as Magento feature packages
+1. Each instance has a dedicated DB (cache, other resources) for tests, so tests can run in isolation
+1. Integration tests must not make network calls to other instances (though any network calls should not be necessary in integration tests)
+
+Changes to integration testing framework:
+1. Application area is configured per instance and does not change per test. `@magentoAppArea` annotation is eliminated as unnecessary.
+1. `\Magento\TestFramework\TestCase\AbstractController` is eliminated. Consider converting tests relying on it (400+) to functional tests.
```

Comment:
I'm not sure if replacing such integration tests by functional tests is a good idea. In some cases, it will be overcomplicated to cover some functionality of a single controller by a functional test because controllers can be entry points for independent business flow. For instance, we have a webhook for Signifyd integration to update case details. To cover such functionality by the functional test we need:
* Go through all checkout flow and place the order
* Go to Signifyd account and make an update for a case
* Go to Admin Panel and verify if the case is updated

Integration test allows doing it much easier because we need only to use fixtures to create an order and stub case and just call a webhook controller with all needed data.

Reply:
As long as an integration test doesn't make network calls to other Magento instances, will it be sufficient?
I still think that Magento instances should be isolated, so running of integration tests doesn't spoil another instance.